If someone wants to reproduce the GCC-16 warning on a fresh install of Fedora 43:

```
dnf install -y git cmake pkgconfig boost-devel libevent-devel sqlite-devel capnproto capnproto-devel zeromq-devel ccache qt6-qttools-devel qt6-qtbase-devel qrencode-devel systemtap-sdt-devel python3-zmq libstdc++

dnf -y copr enable jwakely/gcc-latest && dnf -y install gcc-latest

git clone https://github.com/bitcoin/bitcoin.git  --depth=1 ./b-c && cd b-c

rm -rf ./bld/ && cmake -B bld -DCMAKE_C_COMPILER="/opt/gcc-latest/bin/gcc" -DCMAKE_CXX_COMPILER="/opt/gcc-latest/bin/g++;-static-libstdc++" --preset=dev-mode

cmake --build ./bld -j $(nproc)
```

Will print:

```
[ 17%] Building CXX object src/CMakeFiles/leveldb.dir/leveldb/db/db_impl.cc.o
In copy constructor ‘constexpr leveldb::DBImpl::CompactionState::Output::Output(const leveldb::DBImpl::CompactionState::Output&)’,
    inlined from ‘constexpr _Tp* std::construct_at(_Tp*, _Args&& ...) [with _Tp = leveldb::DBImpl::CompactionState::Output; _Args = {const leveldb::DBImpl::CompactionState::Output&}]’ at /opt/gcc-latest/include/c++/16.0.1/bits/stl_construct.h:110:9,
    inlined from ‘static constexpr void std::allocator_traits<std::allocator<_Up> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = leveldb::DBImpl::CompactionState::Output; _Args = {const leveldb::DBImpl::CompactionState::Output&}; _Tp = leveldb::DBImpl::CompactionState::Output]’ at /opt/gcc-latest/include/c++/16.0.1/bits/alloc_traits.h:676:21,
    inlined from ‘constexpr void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = leveldb::DBImpl::CompactionState::Output; _Alloc = std::allocator<leveldb::DBImpl::CompactionState::Output>]’ at /opt/gcc-latest/include/c++/16.0.1/bits/stl_vector.h:1411:30,
    inlined from ‘leveldb::Status leveldb::DBImpl::OpenCompactionOutputFile(CompactionState*)’ at /b-c/src/leveldb/db/db_impl.cc:808:31:
/b-c/src/leveldb/db/db_impl.cc:57:10: warning: ‘out.leveldb::DBImpl::CompactionState::Output::file_size’ may be used uninitialized [-Wmaybe-uninitialized]
   57 |   struct Output {
      |          ^~~~~~
/b-c/src/leveldb/db/db_impl.cc: In member function ‘leveldb::Status leveldb::DBImpl::OpenCompactionOutputFile(CompactionState*)’:
/b-c/src/leveldb/db/db_impl.cc:804:29: note: ‘out’ declared here
  804 |     CompactionState::Output out;
      |                             ^~~
In copy constructor ‘constexpr leveldb::DBImpl::CompactionState::Output::Output(const leveldb::DBImpl::CompactionState::Output&)’,
    inlined from ‘constexpr _Tp* std::construct_at(_Tp*, _Args&& ...) [with _Tp = leveldb::DBImpl::CompactionState::Output; _Args = {const leveldb::DBImpl::CompactionState::Output&}]’ at /opt/gcc-latest/include/c++/16.0.1/bits/stl_construct.h:110:9,
    inlined from ‘static constexpr void std::allocator_traits<std::allocator<_Up> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = leveldb::DBImpl::CompactionState::Output; _Args = {const leveldb::DBImpl::CompactionState::Output&}; _Tp = leveldb::DBImpl::CompactionState::Output]’ at /opt/gcc-latest/include/c++/16.0.1/bits/alloc_traits.h:676:21,
    inlined from ‘constexpr void std::vector<_Tp, _Alloc>::_M_realloc_append(_Args&& ...) [with _Args = {const leveldb::DBImpl::CompactionState::Output&}; _Tp = leveldb::DBImpl::CompactionState::Output; _Alloc = std::allocator<leveldb::DBImpl::CompactionState::Output>]’ at /opt/gcc-latest/include/c++/16.0.1/bits/vector.tcc:594:26,
    inlined from ‘constexpr void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = leveldb::DBImpl::CompactionState::Output; _Alloc = std::allocator<leveldb::DBImpl::CompactionState::Output>]’ at /opt/gcc-latest/include/c++/16.0.1/bits/stl_vector.h:1417:21,
    inlined from ‘leveldb::Status leveldb::DBImpl::OpenCompactionOutputFile(CompactionState*)’ at /b-c/src/leveldb/db/db_impl.cc:808:31:
/b-c/src/leveldb/db/db_impl.cc:57:10: warning: ‘out.leveldb::DBImpl::CompactionState::Output::file_size’ may be used uninitialized [-Wmaybe-uninitialized]
   57 |   struct Output {
      |          ^~~~~~
/b-c/src/leveldb/db/db_impl.cc: In member function ‘leveldb::Status leveldb::DBImpl::OpenCompactionOutputFile(CompactionState*)’:
/b-c/src/leveldb/db/db_impl.cc:804:29: note: ‘out’ declared here
  804 |     CompactionState::Output out;
      |                             ^~~
[ 17%] Linking CXX static library libleveldb.a
```

(This pull fixes it)

--- a/ci/test/03_test_script.sh
+++ b/ci/test/03_test_script.sh
@@ -44,32 +44,6 @@ echo "=== BEGIN env ==="
 env
 echo "=== END env ==="
 
-# Don't apply patches in the iwyu job, because it relies on the `git diff`
-# command to detect IWYU errors. It is safe to skip this patch in the iwyu job
-# because it doesn't run a UB detector.
-if [[ "${RUN_IWYU}" != true ]]; then
-  # compact->outputs[i].file_size is uninitialized memory, so reading it is UB.
-  # The statistic bytes_written is only used for logging, which is disabled in
-  # CI, so as a temporary minimal fix to work around UB and CI failures, leave
-  # bytes_written unmodified.
-  # See https://github.com/bitcoin/bitcoin/pull/28359#issuecomment-1698694748
-  # Tee patch to stdout to make it clear CI is testing modified code.
-  tee >(patch -p1) <<'EOF'
---- a/src/leveldb/db/db_impl.cc
-+++ b/src/leveldb/db/db_impl.cc
-@@ -1028,9 +1028,6 @@ Status DBImpl::DoCompactionWork(CompactionState* compact) {
-       stats.bytes_read += compact->compaction->input(which, i)->file_size;
-     }
-   }
--  for (size_t i = 0; i < compact->outputs.size(); i++) {
--    stats.bytes_written += compact->outputs[i].file_size;
--  }
-
-   mutex_.Lock();
-   stats_[compact->compaction->level() + 1].Add(stats);
-EOF
-fi
-
 if [ "$RUN_FUZZ_TESTS" = "true" ]; then
   export DIR_FUZZ_IN=${DIR_QA_ASSETS}/fuzz_corpora/
   if [ ! -d "$DIR_FUZZ_IN" ]; then

--- a/db/db_impl.cc
+++ b/db/db_impl.cc
@@ -803,6 +803,7 @@ Status DBImpl::OpenCompactionOutputFile(CompactionState* compact) {
     pending_outputs_.insert(file_number);
     CompactionState::Output out;
     out.number = file_number;
+    out.file_size = 0;
     out.smallest.Clear();
     out.largest.Clear();
     compact->outputs.push_back(out);

